src/lem/circuit.rs

```diff
@@ -56,9 +58,157 @@ use super::{
     Block, Ctrl, Func, Op, Tag, Var,
 };
 
+#[derive(Clone)]
+pub struct MultiFrame<'a, F: LurkField> {
+    pub func: &'a Func,
+    pub store: Option<&'a Store<F>>,
+    pub input: Option<Vec<Ptr<F>>>,
+    pub output: Option<Vec<Ptr<F>>>,
+    pub frames: Option<Vec<Frame<F>>>,
+    pub cached_witness: Option<WitnessCS<F>>,
+    pub count: usize,
+}
+
+impl<'a, F: LurkField> MultiFrame<'a, F> {
+    pub fn blank(func: &'a Func, count: usize) -> Self {
+        Self {
+            func,
+            store: None,
+            input: None,
+            output: None,
+            frames: None,
+            cached_witness: None,
+            count,
+        }
+    }
+
+    pub fn get_store(&self) -> &Store<F> {
+        self.store.expect("store missing")
+    }
+
+    pub fn from_frames(
+        func: &'a Func,
+        count: usize,
+        frames: &[Frame<F>],
+        store: &'a Store<F>,
+    ) -> Vec<Self> {
+        // `count` is the number of `Frames` to include per `MultiFrame`.
+        let total_frames = frames.len();
+        let n = total_frames / count + usize::from(total_frames % count != 0);
+        let mut multi_frames = Vec::with_capacity(n);
+
+        for chunk in frames.chunks(count) {
+            let mut inner_frames = Vec::with_capacity(count);
+
+            for frame in chunk {
+                inner_frames.push(frame.clone());
+            }
+
+            let last_frame = chunk.last().expect("chunk must not be empty");
+            let last_circuit_frame = inner_frames
+                .last()
+                .expect("chunk must not be empty")
+                .clone();
```

Comment:
I think `last_circuit_frame` will always be a clone of `last_frame`. Instead of cloning it for the second time here, even if not needed (then cloning yet again later for every padding frame) — why not skip this assignment completely?